src/modules/flow/selector/selector.c

```diff
@@ -0,0 +1,688 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sol-flow/selector.h"
+#include "sol-flow-internal.h"
+
+#include "sol-util.h"
+#include "sol-buffer.h"
+
+#include <errno.h>
+
+#define MIN_SIZE(a, b) (((a) < (b)) ? (a) : (b))
+#define MAX_SIZE(a, b) (((a) > (b)) ? (a) : (b))
+
+#define DITCH_NL (true)
+#define KEEP_NL (false)
+
+#define DO_FORMAT (false)
+#define CALC_ONLY (true)
+
+static const char CR = '\r', NL = '\n', SPC = ' ', NUL = '\0';
+static const char *TITLE_TAG = "{title}", *VALUE_TAG = "{value}";
+
+struct selector_data {
+    struct sol_flow_node *node;
+    char *title, *sel_mark, *cursor_mark, *text_mem;
+    char *format, *title_tag, *value_tag;
+    struct sol_buffer text_grid;
+    struct sol_ptr_vector items;
+    uint16_t selection, cursor, n_values;
+    size_t columns, rows;
+    bool circular : 1;
+    bool enabled : 1;
+    bool n_values_done : 1;
+};
+
+static inline size_t
+get_buf_size(struct selector_data *mdata)
+{
+    /* +1 on cols for '\n' and (final) NUL chars */
+    return mdata->rows * (mdata->columns + 1);
+}
+
+static inline void
+buffer_re_init(struct selector_data *mdata)
+{
+    /* We choose to do the ending nul byte ourselves, since we're
+     * doing fixed capacity, to able to set the second to last byte on
+     * the buffer without reallocation attempts. */
+    sol_buffer_init_flags(&mdata->text_grid, mdata->text_mem,
+        get_buf_size(mdata), SOL_BUFFER_FLAGS_FIXED_CAPACITY |
+        SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+    memset(mdata->text_mem, 0, sizeof(*mdata->text_mem));
+}
+
+static int
+selector_open(struct sol_flow_node *node,
+    void *data,
+    const struct sol_flow_node_options *options)
+{
+    int r;
+    struct selector_data *mdata = data;
+    const struct sol_flow_node_type_selector_options *opts =
+        (const struct sol_flow_node_type_selector_options *)options;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_SELECTOR_OPTIONS_API_VERSION, -EINVAL);
+
+    mdata->circular = opts->circular;
+    mdata->node = node;
+
+    if (opts->columns.val <= 0) {
+        SOL_WRN("Selector columns number must be a positive integer, "
+            "but %d was given.", opts->columns.val);
+        return -EINVAL;
+    } else
+        mdata->columns = opts->columns.val;
+
+    if (opts->rows.val <= 0) {
+        SOL_WRN("Selector rows number must be a positive integer, "
+            "but %d was given.", opts->rows.val);
+        return -EINVAL;
```

Comment:
Just set rows to "1" and continue. Sure, warning should say that.
We're trying to follow this approach in our node types to be less frustrating as possible to users.